### DISCUSSION TITLE
Using EdgeX for Software Update of End devices

### DISCUSSION THREAD

**Discussion by sudhamani-hcl** [edgexfoundry/community#162] [General]
Hi Team,

I am working on the **software update of end devices** requirement.

The environment is `Cloud --> Gateway (EdgeX)--> End device`. 
**Protocol Support:**
`Cloud --> Gateway`  HTTP or MQTT
`Gateway --> End device`  HTTP or MQTT.

I am running EdgeX in Gateway. In the Gateway, I need to receive software update file from the cloud and forward it to end device.

How can I achieve this using EdgeX?

Thanks in advance,
Sudhamani

**Comment by cloudxxx8** [comment]
It depends on how the End Device API looks like.  The HTTP or MQTT should be enhanced to support.
Cloud --> Gateway can leverage the external MQTT in Core Command service
https://docs.edgexfoundry.org/3.0/microservices/core/command/Ch-Command/#commands-via-messaging

**Reply by lenny-goodell** [reply]
@cloudxxx8 , another good topic for discussion in the next Core WG meeting

**Reply by cloudxxx8** [reply]
this should belong to Device WG topic, but I am not sure what we can do here
@CelineHyz please add to the next Device WG agenda
@lindseysimple @FelixTing awareness

**Comment by sudhamani-hcl** [comment]
Thanks @cloudxxx8 for the info. We'll check this link and come back.

**Comment by lindseysimple** [comment]
Hi @sudhamani-hcl , just FYI we open https://github.com/edgexfoundry/device-sdk-go/issues/1496 to support binary data type in set command in go device sdk and will implement it soon, thanks.
